File: InClassCode/Week2Code/InClassLibraryExample/Library.java
public class Library {
    private Book[] books;
    private int curNumBooks;

    public Library(int capacity) {
        this.books = new Book[capacity];
        this.curNumBooks = 0;
    }
    public boolean addBook(Book book) {
        boolean didAdd = false;
        if(this.curNumBooks < this.books.length && book != null) {
            this.books[this.curNumBooks] = book;
            this.curNumBooks += 1;
            didAdd = true;
        }
        return didAdd;
    }

    public int findBook(Book book) {
        boolean didFind = false;
        int index = -1;

        if(book != null) {
            for(int i=0; i<curNumBooks && !didFind; i+=1) {
                Book curBook = books[i];
                boolean sameTitle = curBook.getTitle().equals(book.getTitle());
                boolean sameAuthor = curBook.getAuthor().equals(book.getAuthor());
                didFind = sameTitle && sameAuthor;
                index = i;
            }
        }
        return index;
    }

    public boolean checkedOut(Book book) {
        int index = findBook(book);
        boolean didCheckout=false;
        if (index >= 0 && !books[index].getCheckedOut()) {
            books[index].changeStatus();
            didCheckout=true;
        }
        return didCheckout;
    }

    public boolean returnBook(Book book){
        int index = findBook(book);
        boolean didReturn=false;
        if (index >= 0 && books[index].getCheckedOut()) {
            books[index].changeStatus();
            didReturn=true;
        }
        return didReturn;
    }
}

File: InClassCode/Week2Code/InClassLibraryExample/library.py
from book import Book

class Library:
    def __init__(self, capacity):
        self._books = [None] * capacity
        self._cur_num_books = 0

    def add_book(self, book):
        did_add = False
        if self._cur_num_books < len(self._books) and book is not None:
            self._books[self._cur_num_books] = book
            self._cur_num_books += 1
            did_add = True
        return did_add

    def find_book(self, book):
        did_find = False
        index = -1

        if book is not None:
            i = 0
            while i < self._cur_num_books and not did_find:
                cur_book = self._books[i]
                if cur_book is not None:
                    same_title = cur_book.get_title() == book.get_title()
                    same_author = cur_book.get_author() == book.get_author()
                    did_find = same_title and same_author

                index = i
                i += 1

        return index

    def checked_out(self, book):
        index = self.find_book(book)
        did_checkout = False
        if index >= 0 and self._books[index] is not None and not self._books[index].get_checked_out():
            self._books[index].change_status()
            did_checkout = True
        return did_checkout

    def return_book(self, book):
        index = self.find_book(book)
        did_return = False
        if index >= 0 and self._books[index] is not None and self._books[index].get_checked_out():
            self._books[index].change_status()
            did_return = True
        return did_return
